src/plugins/intel_cpu/src/memory_state.cpp

```diff
@@ -1,18 +1,131 @@
 // Copyright (C) 2018-2023 Intel Corporation
 // SPDX-License-Identifier: Apache-2.0
 //
+#include <nodes/common/cpu_convert.h>
 
 #include "memory_state.h"
 
 #include "dnnl_extension_utils.h"
+#include "blob_factory.hpp"
+#include "cpu_tensor.h"
 
 using namespace InferenceEngine;
 
 namespace ov {
 namespace intel_cpu {
 
-void VariableState::reset() {
-    std::memset(m_state->data(), 0, m_state->get_byte_size());
+VariableStateDoubleBuffer::VariableStateDoubleBuffer(std::string name,
+                                                     const MemBuilder& mem_build,
+                                                     MemoryDescPtr external_desc,
+                                                     MemoryCPtr init_val) :
+    IVariableState{name}, m_external_desc{external_desc} {
+    reset_prime_mem(mem_build());
+    reset_second_mem(mem_build());
+    m_internal_desc = prime_mem()->getDescPtr();
+    auto&& shape = m_internal_desc->getShape();
+    //TODO what if by some reason we already have internal static state while the node is dynamic, is it even possible?
+
+    if (shape.isStatic()) {
+        if (init_val) {
+            prime_mem()->load(*init_val);
+        } else {
+            prime_mem()->nullify();
+        }
+    } else {
+        //in the case of the original desc has dynamic shape we create an empty tensor
+        auto new_desc = to_static(m_internal_desc);
+        prime_mem()->redefineDesc(new_desc);
+    }
+}
+
+void VariableStateDoubleBuffer::set_state(const ov::SoPtr<ov::ITensor>& state) {
+    m_state = state; // simply to extend the lifetime
+    auto state_desc = MemoryDescUtils::generateCpuBlockedMemoryDesc(m_state);
+
+    const auto& shape = state_desc->getShape();
+
+    if (prime_mem()->getShape() != shape) {
+        auto new_desc = m_internal_desc->cloneWithNewDims(shape.getStaticDims());
+        prime_mem()->redefineDesc(new_desc);
+    }
+
+    auto src = m_state->data();
+
+    Memory mem(get_engine(), state_desc, src);
+    prime_mem()->load(mem);
+}
```

Comment:
Do we need a check to check if `state` is actually `prime_mem()`? If so we can avoid a copy here. e.g., in a user call sequence `state = get_state()` -> do some change in place -> `set_state(state)`.

Reply:
Yep, we can have some optimizations on this level. For example we can try to reuse the user's memory directly as a ro input on the first iteration, but it requires some additional checks to prevent spoiling user's memory. With no doubt it can be optimized later within a separate task.